[DFT] MKLCPU backend does not build with DPC++ release compilers

# Summary
When building with DPC++ 2023.1.0 or 2023.2.0 the following error is encountered:

```
/home/hugh/dev/onemkl/oneMKL/src/dft/backends/mklcpu/forward.cpp:160:27: error: const_cast from 'global_ptr<float>' (aka 'multi_ptr<float, access::address_space::global_space, (sycl::access::decorated)2>') to 'float *' is not allowed
            auto in_ptr = const_cast<input_type *>(in_acc.get_pointer());
                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/hugh/dev/onemkl/oneMKL/src/dft/backends/backend_forward_instantiations.cxx:68:1: note: in instantiation of function template specialization 'oneapi::mkl::dft::mklcpu::compute_forward<oneapi::mkl::dft::detail::descriptor<oneapi::mkl::dft::detail::precision::SINGLE, oneapi::mkl::dft::detail::domain::REAL>, float, std::complex<float>>' requested here
ONEMKL_DFT_FORWARD_INSTANTIATIONS(desc_rf_t, float, float, std::complex<float>)
^
/home/hugh/dev/onemkl/oneMKL/src/dft/backends/backend_forward_instantiations.cxx:36:33: note: expanded from macro 'ONEMKL_DFT_FORWARD_INSTANTIATIONS'
    template ONEMKL_EXPORT void compute_forward<DESCRIPTOR_T, FORWARD_T, BACKWARD_T>(           \
```

This relates to https://github.com/oneapi-src/oneMKL/pull/347, where changes for the nightly Intel LLVM were required.

# Version
oneMKL latest: 6ecec198b275c7b83c93faaae35b6c35950a2203

# Environment

* i9-11900K
* MKLCPU from OneAPI base toolkit 2023.2.0
* Ubuntu 20.02
* DPC++ 2023.1.0 or 2023.2.0
* Error given above

# Steps to reproduce
```
cmake -GNinja ../oneMKL/ -DCMAKE_CXX_COMPILER=icpx -DTARGET_DOMAINS=dft -DENABLE_MKLCPU_BACKEND=ON -DMKL_ROOT=/opt/intel/oneapi/mkl/latest
ninja
```

--- a/src/dft/backends/mklcpu/backward.cpp
+++ b/src/dft/backends/mklcpu/backward.cpp
@@ -156,7 +156,7 @@ ONEMKL_EXPORT void compute_backward(descriptor_type &desc, sycl::buffer<input_ty
         auto out_acc = out.template get_access<sycl::access::mode::write>(cgh);
 
         detail::host_task<class host_kernel_back_outofplace>(cgh, [=]() {
-            auto in_ptr = const_cast<input_type *>(in_acc.get_pointer());
+            auto in_ptr = const_cast<input_type *>(&in_acc.get_pointer()[0]);
             DFT_ERROR status =
                 DftiComputeBackward(desc_acc[detail::DIR::bwd], in_ptr, out_acc.get_pointer());
             if (status != DFTI_NO_ERROR) {
@@ -192,8 +192,8 @@ ONEMKL_EXPORT void compute_backward(descriptor_type &desc, sycl::buffer<input_ty
         auto outim_acc = out_im.template get_access<sycl::access::mode::write>(cgh);
 
         detail::host_task<class host_kernel_split_back_outofplace>(cgh, [=]() {
-            auto inre_ptr = const_cast<input_type *>(inre_acc.get_pointer());
-            auto inim_ptr = const_cast<input_type *>(inim_acc.get_pointer());
+            auto inre_ptr = const_cast<input_type *>(&inre_acc.get_pointer()[0]);
+            auto inim_ptr = const_cast<input_type *>(&inim_acc.get_pointer()[0]);
             DFT_ERROR status =
                 DftiComputeBackward(desc_acc[detail::DIR::bwd], inre_ptr, inim_ptr,
                                     outre_acc.get_pointer(), outim_acc.get_pointer());

--- a/src/dft/backends/mklcpu/forward.cpp
+++ b/src/dft/backends/mklcpu/forward.cpp
@@ -157,7 +157,7 @@ ONEMKL_EXPORT void compute_forward(descriptor_type &desc, sycl::buffer<input_typ
         auto out_acc = out.template get_access<sycl::access::mode::write>(cgh);
 
         detail::host_task<class host_kernel_outofplace>(cgh, [=]() {
-            auto in_ptr = const_cast<input_type *>(in_acc.get_pointer());
+            auto in_ptr = const_cast<input_type *>(&in_acc.get_pointer()[0]);
             DFT_ERROR status =
                 DftiComputeForward(desc_acc[detail::DIR::fwd], in_ptr, out_acc.get_pointer());
             if (status != DFTI_NO_ERROR) {
@@ -193,8 +193,8 @@ ONEMKL_EXPORT void compute_forward(descriptor_type &desc, sycl::buffer<input_typ
         auto outim_acc = out_im.template get_access<sycl::access::mode::write>(cgh);
 
         detail::host_task<class host_kernel_split_outofplace>(cgh, [=]() {
-            auto inre_ptr = const_cast<input_type *>(inre_acc.get_pointer());
-            auto inim_ptr = const_cast<input_type *>(inim_acc.get_pointer());
+            auto inre_ptr = const_cast<input_type *>(&inre_acc.get_pointer()[0]);
+            auto inim_ptr = const_cast<input_type *>(&inim_acc.get_pointer()[0]);
             DFT_ERROR status = DftiComputeForward(desc_acc[detail::DIR::fwd], inre_ptr, inim_ptr,
                                                   outre_acc.get_pointer(), outim_acc.get_pointer());
             if (status != DFTI_NO_ERROR) {

